esp/esp/users/models/__init__.py

```diff
@@ -2549,6 +2549,10 @@ def program_by_perm(cls,user,perm):
         implies = [perm]
         implies+=[x for x,y in cls.implications.items() if perm in y]
 
+        #Check for global permissions (should only work for non-deadlines and admins)
+        if any([cls.user_has_perm(user, x) for x in implies]):
+            return Program.objects.all()
+
```

Comment:
Can you explain why this is the right thing to do?

Reply:
Below this we check for which specific programs the user has the specified permission, but there are a couple permissions that can also be global (e.g. Administrator has every permission globally). Those aren't caught in the code below this, so we need to check if the user has any global permissions that imply the specified permission (`user_has_perm` checks for global permissions when there is no program argument specified).

Reply:
`user_has_perm` says the following.  Do you need to set `program_is_none_implies_all` to True in this call?

`:param program_is_none_implies_all:
            If True, treat Permission objects with program=None as if they are
            global across all programs. Return True if the user has a
            Permission object with program=program or with program=None.
            If False, do not treat Permission objects with program=None as if
            they are global across all programs. Only return True if the user
            has a Permission object with program=program.`

Reply:
I remember getting hung up on that, but I think because `program=None`, then it _only_ checks for global permissions anyways:
```python
:param program:
            Check for permission for `name` on this program.
            If program is None, check only for Permission objects with
            program=None.
            If program_is_none_implies_all is False, check only for Permission
            objects with program=program.
            If program_is_none_implies_all is True, check for Permission
            objects with program=program or program=None.
```

You can see that in `q_permissions_on_program` (which gets called by `user_has_perm`) that the Q filter is already Q(program=program), so the added Q filter when `program_is_none_implies_all=True` wouldn't do anything:
https://github.com/learning-unlimited/ESP-Website/blob/4bc9787dc344d7ea15974ad9386ea98e89fca5d9/esp/esp/users/models/__init__.py#L2405-L2407

Reply:
Ok, got it.